### PR TITLE
Add a troubleshooting guide for common failures in CI services

### PR DESCRIPTION
**Pull request by praneetloke**
### Proposed changes

As the title states, this adds a new page under our Continuous Delivery category to help our users troubleshoot common CI/CD failures when running Pulumi. This guide will be linked to from the Pulumi Console as part of its workflow configuration wizard.

### Unreleased product version (optional)

N/A

### Related issues (optional)

N/A

### CODE REVIEW

File: content/docs/guides/continuous-delivery/troubleshooting-guide.md [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+## Overall Requirements
+
+In order to run a Pulumi command, the following are the requirements for any configuration involving Pulumi:
+
+* A Pulumi Access Token for the account you wish to use.
+Create one [here](https://app.pulumi.com/account/tokens) by logging in with the appropriate account.
+* A stack that you would like to update the automated pipeline.
+* Build tools (more on this below) based on the runtime of your Pulumi app.
+* Restore all dependencies for your Pulumi app.
```
Comment (jkisk) [inline content/docs/guides/continuous-delivery/troubleshooting-guide.md:22]: Perhaps rephrase this to match the rest of the list, which are all things rather than actions?

File: content/docs/guides/continuous-delivery/troubleshooting-guide.md [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+> Learn about [stacks]({{< relref "/docs/intro/concepts/stack" >}}) and their [configuration]({{< relref "/docs/intro/concepts/config" >}}).
+
+A stack represents a specific configuration state for your infrastructure resources. For a typical CI/CD pipeline, the stack must have been created
+beforehand using the `pulumi stack init` command and in the **appropriate organization**.
+
+### Tips
+
+* Ensure that the account represented by the token you are using, has access to the stack.
+  * This can lead to 404s being returned from the Pulumi Service because the token is invalid for any number of reasons.
+* Ensure that you use the fully-qualified stack name when passing the stack name to `pulumi` commands.
```
Comment (jkisk) [inline content/docs/guides/continuous-delivery/troubleshooting-guide.md:50]: perhaps an example  of a fqsn here?

File: content/docs/guides/continuous-delivery/troubleshooting-guide.md [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+* Ensure that you are running the `pulumi` command from the folder containing the `Pulumi.yaml` file that contains your Pulumi project's name.
+* If your stack has configuration, then the stack configuration file must also be co-located with the project file.
+  * For example, for a stack named `production`, the `Pulumi.production.yaml` file must exist alongside the `Pulumi.production.yaml`.
```
Comment (jkisk) [inline content/docs/guides/continuous-delivery/troubleshooting-guide.md:53]: do you mean alongside the pulumi.yaml?

File: content/docs/guides/continuous-delivery/troubleshooting-guide.md [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+  * If your Pulumi app is in a different folder, you can use the `--cwd` flag with almost every `pulumi` command.
+  Learn more about the global flags [here]({{< relref "/docs/reference/cli/#options" >}}).
+
+## Build Tools
+
+Pulumi invokes the build tool that corresponds to your Pulumi project's runtime. If your pipeline is missing one of these tools,
```
Comment (jkisk) [inline content/docs/guides/continuous-delivery/troubleshooting-guide.md:59]: perhaps examples here?

File: content/docs/guides/continuous-delivery/troubleshooting-guide.md [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+when you run `pulumi preview` or `pulumi update --yes`.
+
+### Tips
+
+* For `Node` and `Python` runtimes, you will need to run the step to restore the dependencies ahead of time.
+* For `.NET` and `Go` runtimes, the dependencies are restored for you automatically when you run `pulumi preview` or `pulumi update --yes`.
+* There is an exception to restoring dependencies automatically for `.NET` when you use a private package feed. You must ensure that the
+package(s) from the private feed are accessible or you can use a pre-built binary (see the `binary` option [here]({{< relref "/docs/intro/concepts/project" >}})) with Pulumi to avoid rebuilding your `.NET` solution again.
+  * Note that if you do choose to use a pre-built binary, automatic plugin acquisition for Pulumi
+* You are caching the library dependencies but not the Pulumi plugins. Some services offer dependency caching by captuing a specific folder and restoring
```
Comment (jkisk) [inline content/docs/guides/continuous-delivery/troubleshooting-guide.md:83]: captuing > capturing

File: content/docs/guides/continuous-delivery/troubleshooting-guide.md [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+does not persist across steps, so you cannot split-out the installation step of the Pulumi CLI into another step and expect to use the modified environment
+in the following step. There are exceptions to this. Some CI services allow you to create a reusable template step that you can inject as a pre-cursor to other
+steps. This is a way to reduce repeating the installation step every time you would like to run `pulumi` command.  
+
+### Still need help?
+
+If your error is not related to any of the above categories, it is likely that there is a problem with the code in your infrastructure code.
+Our Programming Model docs can be very helpful in understanding the concepts of programming an infrastructure app. If that doesn't help you solve your
+particular issue, then there is always the power of the strong, knowledgable [Pulumi Slack Community](https://slack.pulumi.com). Signup is free and quick. In addition to other
+community members, various members of the Pulumi team themsevles hang out in the Slack channels and are willing to help you.
```
Comment (jkisk) [inline content/docs/guides/continuous-delivery/troubleshooting-guide.md:122]: themsevles > themselves
maybe "would be happy to help you" instead of "are willing"?

**Review by jkisk — APPROVED** [review]
This looks incredibly helpful!

File: content/docs/guides/continuous-delivery/troubleshooting-guide.md (hunk 1 above)
Comment (praneetloke) [reply content/docs/guides/continuous-delivery/troubleshooting-guide.md:22]: Ah great catch! I'll drop "restore".

File: content/docs/guides/continuous-delivery/troubleshooting-guide.md (hunk 2 above)
Comment (praneetloke) [reply content/docs/guides/continuous-delivery/troubleshooting-guide.md:50]: Oh shoot! I totally meant to do that. Glad you caught this. :)

File: content/docs/guides/continuous-delivery/troubleshooting-guide.md (hunk 3 above)
Comment (praneetloke) [reply content/docs/guides/continuous-delivery/troubleshooting-guide.md:53]: Oops. Yes, I did.

File: content/docs/guides/continuous-delivery/troubleshooting-guide.md (hunk 6 above)
Comment (praneetloke) [reply content/docs/guides/continuous-delivery/troubleshooting-guide.md:122]: Definitely!

File: content/docs/guides/continuous-delivery/troubleshooting-guide.md (hunk 4 above)
Comment (praneetloke) [reply content/docs/guides/continuous-delivery/troubleshooting-guide.md:59]: Done

File: content/docs/guides/continuous-delivery/troubleshooting-guide.md [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+  * If in doubt about problems encountered during execution, clear out all caches and restore dependencies from scratch.
+
+> Note that depending on the number of providers you use in your Pulumi app, the cache size can get big. Some CI services have
+> restrictions on the max size of the cache.
+
+## Cloud Provider Credentials
+
+Another common point of failure is the cloud provider credentials. The class of errors related to the credentials can be, but not limited to:
+
+* Incorrect credentials (wrong account, mismatched keys etc.)
```
Comment (cnunciato) [inline content/docs/guides/continuous-delivery/troubleshooting-guide.md:101]:
```suggestion
* Incorrect credentials (wrong account, mismatched keys, etc.)
```

File: content/docs/guides/continuous-delivery/troubleshooting-guide.md [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+* Keys with a strict access scope that don't have access to creating specific resources
+
+### Tips
+
+* Cloud provider credentials not specified using a recognized mechanism. In almost all cases, specifying the credentials through environment variables will work.
```
Comment (cnunciato) [inline content/docs/guides/continuous-delivery/troubleshooting-guide.md:106]: This doesn't read like a tip. Should it start with a "Make sure" or the like? In general, these should probably all read in the second person: "Make sure you've exposed your provider credentials in a supported way", for example, and ideally with a link to what the supported ways are.

File: content/docs/guides/continuous-delivery/troubleshooting-guide.md [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+
+In order to run a Pulumi command, the following are the requirements for any configuration involving Pulumi:
+
+* A Pulumi Access Token for the account you wish to use.
+Create one [here](https://app.pulumi.com/account/tokens) by logging in with the appropriate account.
+* A stack that you would like to update the automated pipeline.
+* Build tools (more on this below) based on the runtime of your Pulumi app.
+* Dependencies for your Pulumi app.
+* The right cloud provider credentials.
+
```
Comment (cnunciato) [inline content/docs/guides/continuous-delivery/troubleshooting-guide.md:24]: Worth mentioning that Pulumi itself should be installed? Perhaps obvious, but could be a good opportunity to describe or link off to the proper ways of installing Pulumi in CI, mention that Pulumi comes pre-installed on GitHub Actions builders, etc.

File: content/docs/guides/continuous-delivery/troubleshooting-guide.md [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+### Tips
+
+* Cloud provider credentials not specified using a recognized mechanism. In almost all cases, specifying the credentials through environment variables will work.
+  * Some cloud providers also support authentication via their respective CLIs. For example, the Azure provider supports authentication via the Azure CLI.
+  * In the above example, if the Azure CLI is not properly authenticated, then Pulumi will not be able to acquire the necessary credentials.
+* Ensure that the Pulumi step has access to the environment variables containing the credentials.
+  * If your CI service has features that allow you to restrict which pipelines or steps in a pipeline can access the secrets, then ensure that
+  your pipeline/steps can access them.
+* Typo in the name of the environment variables. Refer to our [Cloud Providers]({{< relref "/docs/intro/cloud-providers" >}}) docs and the respective setup pages for each to find the correct
```
Comment (cnunciato) [inline content/docs/guides/continuous-delivery/troubleshooting-guide.md:112]: Similarly, "Check for typos..."

File: content/docs/guides/continuous-delivery/troubleshooting-guide.md [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+### Pulumi not found in the PATH
+
+This problem is especially prevalent for users using a Docker container-based pipeline. With most Docker container pipelines, the state of the container
+does not persist across steps, so you cannot split-out the installation step of the Pulumi CLI into another step and expect to use the modified environment
+in the following step. There are exceptions to this. Some CI services allow you to create a reusable template step that you can inject as a pre-cursor to other
+steps. This is a way to reduce repeating the installation step every time you would like to run `pulumi` command.  
+
+### Still need help?
+
+If your error is not related to any of the above categories, it is likely that there is a problem with the code in your infrastructure code.
```
Comment (cnunciato) [inline content/docs/guides/continuous-delivery/troubleshooting-guide.md:126]: .. or it could be that we just didn't think of it or haven't documented it yet. ;) I do like referring to Slack for this, for sure. I'm not sure I'd point someone into the abyss of the programming model docs to troubleshoot CI problems, though. If they're here, they're likely beyond that level and looking for CI-specific info.

**Review by cnunciato — COMMENTED** [review]
Buncha little things for ya to consider.

File: content/docs/guides/continuous-delivery/troubleshooting-guide.md (hunk 9 above)
Comment (praneetloke) [reply content/docs/guides/continuous-delivery/troubleshooting-guide.md:24]: Yes, absolutely! Completely forgot that.

File: content/docs/guides/continuous-delivery/troubleshooting-guide.md [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+
+A stack represents a specific configuration state for your infrastructure resources. For a typical CI pipeline, the stack must have been created
+beforehand using the `pulumi stack init` command and in the **appropriate organization**.
+
+### Tips
+
+* Ensure that the account represented by the token you are using, has access to the stack.
+  * This can lead to 404s being returned from the Pulumi Service because the token is invalid for any number of reasons.
+* Ensure that you use the fully-qualified stack name when passing the stack name to `pulumi` commands.
+  * A fully-qualified stack name is of the format `<org_name>/<project_name>/<stack_name>`.
```
[the remark this answers is not in the document]
Comment (praneetloke) [reply content/docs/guides/continuous-delivery/troubleshooting-guide.md:51]: Yes, very much agree with that.

File: content/docs/guides/continuous-delivery/troubleshooting-guide.md [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+  * For example, for a stack called `production` in an org called `pulumi` and project `slack-bot`, its FQDN is `pulumi/slack-bot/production`.
+  * Using an FQDN in your automated pipelines removes any ambiguity as to which stack is being used, even though Pulumi is smart to figure out
+  which one you meant based on the project name in your `Pulumi.yaml` file. It makes it explicit for anyone in your team looking at the CI configuration.
+* Ensure that you are running the `pulumi` command from the folder containing the `Pulumi.yaml` file that contains your Pulumi project's name.
+* If your stack has configuration, then the stack configuration file must also be co-located with the project file.
+  * For example, for a stack named `production`, the `Pulumi.production.yaml` file must exist alongside the `Pulumi.yaml`.
```
[the remark this answers is not in the document]
Comment (praneetloke) [reply content/docs/guides/continuous-delivery/troubleshooting-guide.md:57]: > Technically it just has to exist at or below the dir holding Pulumi.yaml (e.g., for those who put their stack configs in a subfolder).

Oh! I didn't know that myself! Hmm is there an example of this somewhere?

File: content/docs/guides/continuous-delivery/troubleshooting-guide.md [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+
+## Restoring Dependencies
+
+Regardless of the language you pick for your Pulumi app, you will have dependencies on at least one other package or library. You must ensure that you
+restore these dependencies before running a `pulumi` command. In the case of `.NET` and `Go` runtimes, those dependencies are automatically restored for you
+when you run `pulumi preview` or `pulumi update --yes`.
+
+### Tips
+
+* For `Node` and `Python` runtimes, you will need to run the step to restore the dependencies ahead of time.
```
[the remark this answers is not in the document]
Comment (praneetloke) [reply content/docs/guides/continuous-delivery/troubleshooting-guide.md:83]: I will rephrase this sentence.

File: content/docs/guides/continuous-delivery/troubleshooting-guide.md [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+* For `.NET` and `Go` runtimes, the dependencies are restored for you automatically when you run `pulumi preview` or `pulumi update --yes`.
+* There is an exception to restoring dependencies automatically for `.NET` when you use a private package feed. You must ensure that the
+package(s) from the private feed are accessible or you can use a pre-built binary (see the `binary` option [here]({{< relref "/docs/intro/concepts/project" >}})) with Pulumi to avoid rebuilding your `.NET` solution again.
+  * Note that if you do choose to use a pre-built binary, automatic plugin acquisition for Pulumi
```
[the remark this answers is not in the document]
Comment (praneetloke) [reply content/docs/guides/continuous-delivery/troubleshooting-guide.md:87]: Yeah I just noticed that too myself while I was doing something else and fixed it now.

File: content/docs/guides/continuous-delivery/troubleshooting-guide.md [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+* You are caching the library dependencies but not the Pulumi plugins. Some services offer dependency caching by capturing a specific folder and restoring
```
[the remark this answers is not in the document]
Comment (praneetloke) [reply content/docs/guides/continuous-delivery/troubleshooting-guide.md:88]: It's the latter, but I'll think of a different way to phrase this.

File: content/docs/guides/continuous-delivery/troubleshooting-guide.md [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+that folder whenever you run your pipeline subsequently. However, note that Pulumi dependencies have a post-install step that also pulls-down
+a [plugin]({{< relref "/docs/intro/concepts/how-pulumi-works#resource-providers" >}}) binary from our CDN.
+  * So be sure to cache the plugins path as well.
```
[the remark this answers is not in the document]
Comment (praneetloke) [reply content/docs/guides/continuous-delivery/troubleshooting-guide.md:91]: Yes, agreed. I have fixed this.

File: content/docs/guides/continuous-delivery/troubleshooting-guide.md (hunk 8 above)
Comment (praneetloke) [reply content/docs/guides/continuous-delivery/troubleshooting-guide.md:106]: I will rephrase this sentence to be similar to what you suggest.

File: content/docs/guides/continuous-delivery/troubleshooting-guide.md [hunk 18]
```diff
@@ -0,0 +1,129 @@
...
+Another common point of failure is the cloud provider credentials. The class of errors related to the credentials can be, but not limited to:
+
+* Incorrect credentials (wrong account, mismatched keys etc.)
+* Keys with a strict access scope that don't have access to creating specific resources
+
+### Tips
+
+* Cloud provider credentials not specified using a recognized mechanism. In almost all cases, specifying the credentials through environment variables will work.
+  * Some cloud providers also support authentication via their respective CLIs. For example, the Azure provider supports authentication via the Azure CLI.
+  * In the above example, if the Azure CLI is not properly authenticated, then Pulumi will not be able to acquire the necessary credentials.
```
[the remark this answers is not in the document]
Comment (praneetloke) [reply content/docs/guides/continuous-delivery/troubleshooting-guide.md:108]: Ah yeah. Fixed it.

File: content/docs/guides/continuous-delivery/troubleshooting-guide.md (hunk 10 above)
Comment (praneetloke) [reply content/docs/guides/continuous-delivery/troubleshooting-guide.md:112]: Done.

File: content/docs/guides/continuous-delivery/troubleshooting-guide.md (hunk 11 above)
Comment (praneetloke) [reply content/docs/guides/continuous-delivery/troubleshooting-guide.md:126]: Yes, very much agree with this. Reading this again, it actually doesn't sound nice at all. It seems to blame the user's code with high confidence. I will remove all of this and simply suggest to the user to reach out on our Community Slack.

File: content/docs/guides/continuous-delivery/troubleshooting-guide.md (hunk 9 above)
Comment (praneetloke) [reply content/docs/guides/continuous-delivery/troubleshooting-guide.md:25]: I added an item here about the Pulumi CLI, as well as a section about it.

**Comment by praneetloke** [comment]
I'll go ahead and merge this. I addressed all of the comments. Thank you, @jkisk and @cnunciato for your time reviewing this.  The guide looks much better now!

File: content/docs/guides/continuous-delivery/troubleshooting-guide.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (cnunciato) [reply content/docs/guides/continuous-delivery/troubleshooting-guide.md]: I don't know of a public example, but the chapter I'm working on covers it, which is the only reason I happened to think of it myself. 😄 For example, if your Pulumi.yaml looked like this:

```
name: some-app
description: Some description
main: program
config: stacks
```

... the corresponding file structure would look like this:

```
$ tree
.
├── Pulumi.yaml
├── program 
│   ├── node_modules
│   │   └── ...
│   ├── tsconfig.json
│   ├── index.ts
│   ├── package-lock.json
│   ├── package.json
│   └── tsconfig.json
└── stacks
    ├── Pulumi.dev.yaml
    └── Pulumi.prod.yaml
```
